coap-gateway/service/exchange.go

```diff
@@ -0,0 +1,100 @@
+package service
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/plgd-dev/cloud/pkg/log"
+	"github.com/plgd-dev/cloud/pkg/security/oauth2"
+	"github.com/plgd-dev/cloud/pkg/sync/task/future"
+)
+
+type exchangeData struct {
+	token  string
+	future *future.Future
+}
+
+type exchangeHandler struct {
+	lock       sync.Mutex
+	codeToData map[string]*exchangeData
+}
+
+func NewExchangeHandler() *exchangeHandler {
+	return &exchangeHandler{
+		codeToData: make(map[string]*exchangeData),
+	}
+}
+
+func (e *exchangeHandler) removeData(code string) {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+	delete(e.codeToData, code)
+}
+
+func (e *exchangeHandler) setTokenForCode(code, token string) error {
+	if code == "" {
+		return fmt.Errorf("invalid code")
+	}
+	e.lock.Lock()
+	defer e.lock.Unlock()
+	data, ok := e.codeToData[code]
+	if !ok {
+		return fmt.Errorf("item not found")
+	}
+	data.token = token
+	return nil
+}
+
+func (e *exchangeHandler) Exchange(ctx context.Context, provider *oauth2.PlgdProvider, authorizationCode string) (*oauth2.Token, error) {
+	var setToken future.SetFunc
+	var setError future.SetErrorFunc
+	e.lock.Lock()
+	data, ok := e.codeToData[authorizationCode]
+	if !ok {
+		var fut *future.Future
+		fut, setToken, setError = future.New()
+		data = &exchangeData{
+			future: fut,
+		}
+		e.codeToData[authorizationCode] = data
+	}
+	e.lock.Unlock()
```

Comment:
I prefer to use `defer e.lock.Unlock()` so I suggest moving the whole code under lock to a separate function